### PR TITLE
feat(autoconfig): emit cloudflare.config.ts for new Vite projects

### PR DESCRIPTION
**Pull request by penalosa**
Adds **experimental** support to `@cloudflare/autoconfig` for emitting the new programmatic config format when auto-configuring a **new Vite-based project**, gated behind a new opt-in option. Default behaviour is unchanged.

- `runAutoConfig` accepts a new `experimentalConfigFormat` option (`"jsonc"` | `"ts"`, defaulting to `"jsonc"`). When `"ts"` **and the project is Vite-based**, autoconfig writes a `cloudflare.config.ts` (runtime config via `defineWorker`, imported from `@cloudflare/vite-plugin`) instead of `wrangler.jsonc`, and drives the project with `cf`.
- Tooling settings are owned by Vite, so they are surfaced as a warning rather than written to a config file.
- Any `wrangler.jsonc` a framework writes during setup is left **untouched** (no read/merge/delete), since it may not be compatible with the new format. The generated `cloudflare.config.ts` is built purely from the config autoconfig assembles in memory.
- Non-Vite projects (and the default `"jsonc"` format) continue to write `wrangler.jsonc`.

Scope note: this PR deliberately does **not** include migration of existing `wrangler.jsonc` projects, nor a non-Vite `wrangler.config.ts` (tooling) output — those are out of scope.

This is an experimental, internal-only capability. It is fully opt-in: the new behaviour only activates when a caller passes `experimentalConfigFormat: "ts"` for a Vite project, so existing consumers (e.g. `wrangler setup` / `wrangler deploy`) are unaffected and continue to emit `wrangler.jsonc`.

---

- Tests
  - [x] Tests included/updated
  - [ ] Automated tests not possible - manual testing has been completed as follows:
  - [ ] Additional testing not necessary because:
- Public documentation
  - [ ] Cloudflare docs PR(s): 
  - [x] Documentation not necessary because: this is an experimental, internal-only capability with no public-facing API surface yet.

### CODE REVIEW

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>

- .changeset/autoconfig-ts-config-output.md: [@cloudflare/wrangler]
- packages/autoconfig/src/config-module/convert.ts: [@cloudflare/wrangler]
- packages/autoconfig/src/config-module/fs-utils.ts: [@cloudflare/wrangler]
- packages/autoconfig/src/config-module/serialize.ts: [@cloudflare/wrangler]
- packages/autoconfig/src/run.ts: [@cloudflare/wrangler]
- packages/autoconfig/src/types.ts: [@cloudflare/wrangler]
- packages/autoconfig/tests/config-module/convert.test.ts: [@cloudflare/wrangler]
- packages/autoconfig/tests/config-module/serialize.test.ts: [@cloudflare/wrangler]
- packages/autoconfig/tests/run-new-config.test.ts: [@cloudflare/wrangler]
- pnpm-lock.yaml: [@cloudflare/wrangler]

</details>

**Comment by penalosa** [comment]
Addressed the Devin review comments in 8a0d88275:

- 🔴 **Overly aggressive regex in script rewriting** (`migrate.ts`) — Fixed. The `/\bwrangler\b/g` pattern is now `/(^|[\s;&|])wrangler(?=\s|$)/g`, which only matches `wrangler` as a standalone command token (start of script or after a shell separator, followed by whitespace/end). File references like `wrangler.jsonc` and compound names like `@cloudflare/wrangler-foo` / `wrangler-pages` are left untouched. Added a test covering these edge cases.

- 🟡 **`deployCommand`/`versionCommand` not updated for `ts` format** (`run.ts`) — Fixed. In `ts` mode the default deploy/version commands are now `cf deploy` / `cf versions upload`, consistent with the generated `cf` scripts (framework overrides still take precedence). Test asserts the summary commands.

- 🟡 **Framework-created `wrangler.jsonc` not cleaned up in `ts` mode** (`run.ts`) — Fixed. `saveNewConfig` now falls back to re-discovering the current config path via `getWranglerJsonConfigPath`, so a `wrangler.jsonc` written by the framework's `configure()` is removed too (not just a pre-existing one). Added a test with a framework that writes its own config.

- 🚩 **`migrate.ts` install doesn't pass `isWorkspaceRoot`** — Fixed. Added an optional `isWorkspaceRoot` to `MigrateOptions` (default `false`) and forward it to `installPackages` so `cf` lands in the correct `node_modules` in monorepos. Added a test.

- 🚩 **`convertExports` ignores renames/deletes** (`config/split.ts`) — Fixed. Migrations are now replayed in order: `renamed_classes` move the entry (preserving storage) and `deleted_classes` drop it, so `exports` reflects the current class set rather than every class ever declared. (`transferred_classes` isn't part of the `DurableObjectMigration` type, so there's nothing to handle there.) Added tests for rename + delete.

- 🚩 **`isVite` snapshot captured before `configure()`** (`run.ts`) — Intentionally left as-is. This mirrors the deliberate early capture of `preexistingWranglerConfigPath`: we snapshot before the framework's CLI runs so a config it writes during setup isn't mistaken for pre-existing user state. As the comment notes, Vite-based frameworks already have a `vite.config` before autoconfig runs, so this isn't a practical concern. Reworking it would require moving the install step after `configure()`, which isn't warranted for this experimental, internal-only path.

File: packages/autoconfig/src/config-module/fs-utils.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+export function hasViteConfig(projectPath: string): boolean {
+	return ["ts", "js", "mjs", "mts", "cjs", "cts"].some((ext) =>
+		existsSync(resolve(projectPath, `vite.config.${ext}`))
+	);
+}
+
+/**
+ * Detect the project's package manager from its lockfile, defaulting to npm.
+ */
+export function detectPackageManager(
```
Comment (penalosa) [inline packages/autoconfig/src/config-module/fs-utils.ts]: Is this new logic in this PR, or are you just moving it around?

File: packages/autoconfig/src/config-module/serialize.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ */
+import type { NewConfigSplit } from "@cloudflare/config";
+
+/**
+ * The module specifier the generated files import their `define*` helpers
+ * from. `wrangler/experimental-config` re-exports `defineWorker` (from
+ * `@cloudflare/config`) and `defineWranglerConfig`, so a single dependency
+ * (wrangler, which autoconfig already installs) covers both files.
+ */
+export const CONFIG_IMPORT_SOURCE = "wrangler/experimental-config";
```
Comment (penalosa) [inline packages/autoconfig/src/config-module/serialize.ts]: This should be imported from Vite or Wrangler, depending on whether a wrangler.config.ts file is being created

File: packages/autoconfig/src/config-module/serialize.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+/** Print an object key, quoting it when it isn't a bare identifier. */
+function printKey(key: string): string {
+	return VALID_IDENTIFIER.test(key) ? key : JSON.stringify(key);
+}
+
+/**
+ * Print a JS value as TypeScript source. `indent` is the current indentation
+ * (tabs). Objects and arrays are printed multi-line; empty ones inline.
+ */
+function printValue(value: unknown, indent: string): string {
```
Comment (penalosa) [inline packages/autoconfig/src/config-module/serialize.ts]: Surely a lot of this could just be `JSON.stringify(value, null, indent)`?

File: packages/autoconfig/src/run.ts [hunk 4]
```diff
@@ -358,34 +497,52 @@ export async function buildOperationsSummary(
...
-		) {
-			summary.scripts["cf-typegen"] =
-				packageJsonScriptsOverrides?.typegen ?? "wrangler types";
+		if (isTs) {
+			// `cf deploy` / `cf dev` run the framework build themselves, so the
+			// scripts don't prepend the build command (unlike the wrangler path).
+			summary.scripts = {
+				deploy: "cf deploy",
+				preview: "cf dev",
+			};
```
[the remark this answers is not in the document]
Comment (penalosa) [reply packages/autoconfig/src/run.ts]: This is intentional for cf

File: packages/autoconfig/src/run.ts [hunk 5]
```diff
@@ -310,6 +375,72 @@ async function saveWranglerJsonc(
...
+	// Migration: remove the old wrangler config now that the new files exist.
+	// `preexistingWranglerConfigPath` is captured before the framework's
+	// `configure()` runs; a framework CLI may also write a fresh
+	// `wrangler.jsonc` during setup, so fall back to re-discovering the current
+	// path to ensure that is cleaned up too.
+	const wranglerConfigToRemove =
+		options.preexistingWranglerConfigPath ??
+		getWranglerJsonConfigPath(projectPath);
+	if (wranglerConfigToRemove) {
+		await rm(wranglerConfigToRemove, { force: true });
```
Comment (penalosa) [inline packages/autoconfig/src/run.ts]: Why force:true?

File: packages/config/src/split.ts [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+		worker.env = env;
+	}
+	if (triggers.length > 0) {
+		worker.triggers = triggers;
+	}
+
+	convertTooling(raw, tooling);
+
+	return { worker, tooling };
+}
```
[the remark this answers is not in the document]
Comment (penalosa) [reply packages/config/src/split.ts]: We now warn about these: https://github.com/cloudflare/workers-sdk/pull/14359/commits/75b0d91fbda0fcfabb61e5489017f376ef2bc880

**Comment by penalosa** [comment]
Addressed the review feedback in 735d4fdcf:

- **`fs-utils.ts` `detectPackageManager` — "new logic or moved?"**: It's new (added in this PR for the standalone migrate flow). Reworked it to derive lockfile names from the shared `PackageManager` definitions in `@cloudflare/workers-utils` rather than hardcoding them, so there's a single source of truth.

- **`serialize.ts` import source should be Vite or Wrangler**: `cloudflare.config.ts` now imports `defineWorker` from `@cloudflare/vite-plugin/experimental-config` for Vite projects and `wrangler/experimental-config` otherwise (via a new `cloudflareConfigImportSource(isVite)` helper). `wrangler.config.ts` is only emitted for non-Vite projects, so it always imports from wrangler.

- **`serialize.ts` `printValue` could just be `JSON.stringify`**: Done — the bespoke serializer is gone; config is emitted with plain `JSON.stringify(config, null, "\t")` (quoted keys are fine).

- **`run.ts` `rm(..., { force: true })` — "why force:true?"**: Removed; removals are now guarded by an `existsSync` check instead.

- **Framework-written on-disk config (Devin 🔴/🟡)**: `saveNewConfig` now reads and merges any `wrangler.jsonc` the framework's `configure()` writes to disk (mirroring the jsonc path), so settings from frameworks like Next.js via `@opennextjs/cloudflare` aren't lost, and it removes *all* superseded config files (pre-existing + framework-created), not just one.

Also updated the PR description to follow the template (checkboxes + removed the stale `Fixes #` line).

**Comment by penalosa** [comment]
Narrowed the scope of this PR (ac3a03e83). It is now **only** about emitting `cloudflare.config.ts` instead of `wrangler.jsonc` when auto-configuring a new Vite-based project.

Removed from scope:
- **Migration** — dropped the standalone `migrateWranglerConfigToNewFormat` flow and its tests entirely.
- **Non-Vite TS output** — removed the `wrangler.config.ts` (tooling) output and the build-tool-preservation logic. Non-Vite projects (and the default `jsonc` format) keep writing `wrangler.jsonc`; if `ts` is requested for a non-Vite project it now falls back to `wrangler.jsonc`.
- **Touching framework configs** — `saveNewConfig` no longer reads, merges, or deletes any on-disk `wrangler.jsonc`. A `wrangler.jsonc` written by a framework's `configure()` step is left exactly as-is (it may not be compatible with the new format); the generated `cloudflare.config.ts` is built purely from the config autoconfig assembles in memory.

`cloudflare.config.ts` now always imports `defineWorker` from `@cloudflare/vite-plugin` (the Vite project's build tool). Net diff is ~785 lines smaller. Tests: autoconfig 181, config 123, wrangler autoconfig 27 — all green; lint/format clean.

File: packages/autoconfig/src/config-module/convert.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+		set(assets, "htmlHandling", raw.assets.html_handling);
+		set(assets, "notFoundHandling", raw.assets.not_found_handling);
+		set(assets, "runWorkerFirst", raw.assets.run_worker_first);
+		if (Object.keys(assets).length > 0) {
+			worker.assets = assets;
+		}
+		if (raw.assets.binding) {
+			worker.env = { [raw.assets.binding]: { type: "assets" } };
+		}
+		set(tooling, "assetsDirectory", raw.assets.directory);
```
Comment (jamesopstad) [inline packages/autoconfig/src/config-module/convert.ts:70]: Does autoconfig set the assets directory for Vite projects?

File: packages/autoconfig/src/config-module/convert.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+ * autoconfig only ever writes config for a brand-new project, so binding,
+ * trigger, Durable Object migration, etc. fields never appear here.
+ */
+import type { RawConfig } from "@cloudflare/workers-utils";
+
+type Obj = Record<string, unknown>;
+
+export interface CloudflareConfigSplit {
+	/** Runtime config → `cloudflare.config.ts` (`defineWorker`). */
+	worker: Obj;
```
Comment (jamesopstad) [inline packages/autoconfig/src/config-module/convert.ts:20]: Can this use the `UserConfig` type from `@cloudflare/config`?

File: packages/autoconfig/src/config-module/serialize.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * Serialize the runtime config to `cloudflare.config.ts` source, importing
+ * `defineWorker` from the Cloudflare Vite plugin.
+ */
+export function serializeCloudflareConfig(
+	worker: CloudflareConfigSplit["worker"]
+): string {
+	return (
+		`import { defineWorker } from "${CLOUDFLARE_CONFIG_IMPORT_SOURCE}";\n\n` +
+		`export default defineWorker(${JSON.stringify(worker, null, "\t")});\n`
```
Comment (jamesopstad) [inline packages/autoconfig/src/config-module/serialize.ts:28]: I think we should be using the `bindings` helper (if there is a binding) and the `{ type: "cf-worker" }` `entrypoint` here. I know it makes it a little more complicated to set up but this gives users the best start.

File: packages/autoconfig/src/run.ts [hunk 10]
```diff
@@ -358,34 +458,49 @@ export async function buildOperationsSummary(
...
-			!("cf-typegen" in (autoConfigDetails.packageJson.scripts ?? {}))
-		) {
-			summary.scripts["cf-typegen"] =
-				packageJsonScriptsOverrides?.typegen ?? "wrangler types";
+		if (useNewConfig) {
+			// `cf deploy` / `cf dev` run the framework build themselves, so the
+			// scripts don't prepend the build command (unlike the wrangler path).
+			summary.scripts = {
+				deploy: "cf deploy",
+				preview: "cf dev",
```
Comment (jamesopstad) [inline packages/autoconfig/src/run.ts:475]: This shouldn't be called `preview`. Is the issue here that we shouldn't overwrite the framework's `dev` command?

File: packages/autoconfig/src/types.ts [hunk 11]
```diff
@@ -65,6 +65,16 @@ export type AutoConfigOptions = {
...
 	enableWranglerInstallation?: boolean;
+	/**
+	 * Experimental: which config format to emit (default: `"jsonc"`).
+	 *
+	 * - `"jsonc"` — write a `wrangler.jsonc` (the established behaviour).
+	 * - `"ts"` — write the new programmatic format: a `cloudflare.config.ts`
+	 *   (runtime config via `defineWorker`) and, when migrating an existing
+	 *   non-Vite Wrangler project, a `wrangler.config.ts` (tooling config via
+	 *   `defineWranglerConfig`). Any pre-existing `wrangler.jsonc` is removed.
+	 */
```
Comment (jamesopstad) [inline packages/autoconfig/src/types.ts:76]:
```suggestion
	 * - `"jsonc"` — write a `wrangler.jsonc` (the established behaviour).
	 * - `"ts"` — write the new programmatic format: a `cloudflare.config.ts`
	 *   (runtime config via `defineWorker`).
	 */
```

**Review by jamesopstad — COMMENTED** [review]
Left some comments. Do we also have the issue here that not all frameworks that are detected as using Vite will be compatible with the Cloudflare plugin?
